srtcore/core.cpp

```diff
@@ -4624,24 +4619,15 @@ int CUDT::receiveBuffer(char* data, int len)
                 while (stillConnected() && !m_pRcvBuffer->isRcvDataReady())
                 {
                     //Do not block forever, check connection status each 1 sec.
-                    uint64_t exptime = CTimer::getTime() + 1000000ULL;
-                    timespec locktime;
-
-                    locktime.tv_sec = exptime / 1000000;
-                    locktime.tv_nsec = (exptime % 1000000) * 1000;
-                    pthread_cond_timedwait(&m_RecvDataCond, &m_RecvLock, &locktime);
+                    CTimer::condTimedWaitUS(&m_RecvDataCond, &m_RecvLock, 1000000ULL);
                 }
             }
             else
             {
                 uint64_t exptime = CTimer::getTime() + m_iRcvTimeOut * 1000;
-                timespec locktime;
-                locktime.tv_sec = exptime / 1000000;
-                locktime.tv_nsec = (exptime % 1000000) * 1000;
-
                 while (stillConnected() && !m_pRcvBuffer->isRcvDataReady())
                 {
-                    pthread_cond_timedwait(&m_RecvDataCond, &m_RecvLock, &locktime);
+                    CTimer::condTimedWaitUS(&m_RecvDataCond, &m_RecvLock, m_iRcvTimeOut * 1000);
```

Comment:
Excuse me, no one has noticed this?

The old code precalculates the latest time up to which it should wait, using `CTimer::getTime` once. Then in a loop there's waiting up to this time at most.

The new code makes waiting for `m_iRcvTimeOut*1000` basing on `CTimer::getTime` called again every time in a loop, which makes it wait for exactly that amount of time, regardless of how much time was already spent in previous iterations.

It is not the same.

Reply:
Indeed, with `condTimedWaitUS()` we need to calculate how much time is left to sleep instead of sleeping fixed amount of time. I'm apologize for this fault. Should I wash away my sin, or you'll kindly fix it?